core/src/main/java/amino/run/kernel/client/KernelClient.java

```diff
@@ -28,41 +34,180 @@
  * @author iyzhang
  */
 public class KernelClient {
+    private static final int MIN_METRIC_POLL_PERIOD_MS = 1000; /* Minimum metrics poll period */
+    private static final int MAX_METRIC_POLL_PERIOD_MS = 512000; /* Maximum metrics poll period */
+    private static final int STEP_SIZE_IN_BYTES = 1024; /* Step size in bytes */
+    /* Random byte array used to send in heartbeats */
+    private static final byte[] randomBytes = new byte[STEP_SIZE_IN_BYTES];
+
+    static {
+        new Random().nextBytes(randomBytes);
+    }
+
+    /* Data holder class used in heartbeats */
+    public static class RandomData implements Serializable {
+        private transient int len = STEP_SIZE_IN_BYTES; /* length of data to be sent to server */
```

Comment:
Why do you bother having both len and STEP_SIZE_IN_BYTES?

Don't you mean for the two values to be different?  So that you can send multiples of STEP_SIZE_IN_BYTES?

Also, your logic as it currently stands only supports len values which are strictly multiples of STEP_SIZEIN_BYTES?  

This is allo quite confusing code, which needs to be fixed to work correctly, or properly commented to make it clear.

Reply:
Initial value of `len` is set with `RANDOM_BYTE_LEN`.  During heartbeats, when data size is too small relative to line latency, `len` is doubled.